code/implementation/vm/src/main/java/io/cattle/platform/vm/process/VirtualMachinePreCreate.java

```diff
@@ -0,0 +1,91 @@
+package io.cattle.platform.vm.process;
+
+import io.cattle.platform.core.constants.InstanceConstants;
+import io.cattle.platform.core.model.Instance;
+import io.cattle.platform.core.util.SystemLabels;
+import io.cattle.platform.docker.constants.DockerInstanceConstants;
+import io.cattle.platform.engine.handler.HandlerResult;
+import io.cattle.platform.engine.handler.ProcessPreListener;
+import io.cattle.platform.engine.process.ProcessInstance;
+import io.cattle.platform.engine.process.ProcessState;
+import io.cattle.platform.object.meta.ObjectMetaDataManager;
+import io.cattle.platform.object.util.DataAccessor;
+import io.cattle.platform.process.common.handler.AbstractObjectProcessLogic;
+import io.cattle.platform.util.type.Priority;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class VirtualMachinePreCreate extends AbstractObjectProcessLogic implements ProcessPreListener, Priority {
+
+    private static final String[] CAPS = new String[] { "NET_ADMIN" };
+    private static final String[] VOLUMES = new String[] { "/var/lib/rancher/vm:/vm" };
+    private static final String[] DEVICES = new String[] { "/dev/kvm:/dev/kvm", "/dev/net/tun:/dev/net/tun" };
+
+    @Override
+    public String[] getProcessNames() {
+        return new String[] { "instance.create" };
+    }
+
+    @Override
+    public HandlerResult handle(ProcessState state, ProcessInstance process) {
+        Instance instance = (Instance)state.getResource();
+        if (!InstanceConstants.KIND_VIRTUAL_MACHINE.equals(instance.getKind())) {
+            return null;
+        }
+
+        Map<Object, Object> fields = new HashMap<>();
+        Map<String, Object> labels = DataAccessor.fieldMap(instance, InstanceConstants.FIELD_LABELS);
+
+        labels.put(SystemLabels.LABEL_VM, "true");
+        if (instance.getMemoryMb() == null) {
+            labels.put(SystemLabels.LABEL_VM_MEMORY, "512");
+        } else {
+            labels.put(SystemLabels.LABEL_VM_MEMORY, instance.getMemoryMb().toString());
+        }
+        labels.put(SystemLabels.LABEL_VM_USERDATA, instance.getUserdata());
+        labels.put(SystemLabels.LABEL_VM_VCPU, DataAccessor.fieldString(instance, InstanceConstants.FIELD_VCPU));
+
+        fields.put(InstanceConstants.FIELD_LABELS, labels);
+        fields.put(ObjectMetaDataManager.CAPABILITIES_FIELD, Arrays.asList("console"));
+
+        // TODO: remove this
```

Comment:
@yasker Needs to make a change to read meta data.  I'm fine with merge this as is.  And removing later.